go/mcap/new_reader.go

```diff
@@ -0,0 +1,503 @@
+	channels          map[uint16]*Channel
+	schemas           map[uint16]*Schema
+	chunkIndices      []*ChunkIndex
+	attachmentIndices []*AttachmentIndex
+	metadataIndices   []*MetadataIndex
+}
+
+func parseSummary(rs io.ReadSeeker) (*summary, error) {
+	s := &summary{
+		channels: make(map[uint16]*Channel),
+		schemas:  make(map[uint16]*Schema),
+	}
+	_, err := rs.Seek(-8-4-8-8, io.SeekEnd) // magic, plus 20 bytes footer
+	if err != nil {
+		return nil, err
+	}
+	buf := make([]byte, 8+20)
+	_, err = io.ReadFull(rs, buf)
+	if err != nil {
+		return nil, fmt.Errorf("read error: %w", err)
+	}
+	magic := buf[20:]
+	if !bytes.Equal(magic, Magic) {
+		return nil, fmt.Errorf("not an mcap file")
+	}
+	footer, err := ParseFooter(buf[:20])
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse footer: %w", err)
+	}
+
+	// scan the whole summary section
+	if footer.SummaryStart == 0 {
+		return nil, nil
+	}
+	_, err = rs.Seek(int64(footer.SummaryStart), io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("failed to seek to summary start")
+	}
+	lexer, err := NewLexer(rs, &LexerOptions{SkipMagic: true})
+	if err != nil {
+		return nil, err
+	}
+	for {
+		tokenType, record, err := lexer.Next(nil)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil, nil
+			}
+			return nil, fmt.Errorf("failed to get next token: %w", err)
+		}
+		switch tokenType {
+		case TokenSchema:
+			schema, err := ParseSchema(record)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse schema: %w", err)
+			}
+			s.schemas[schema.ID] = schema
+		case TokenChannel:
+			channel, err := ParseChannel(record)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse channel info: %w", err)
+			}
+			s.channels[channel.ID] = channel
+		case TokenAttachmentIndex:
+			idx, err := ParseAttachmentIndex(record)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse attachment info: %w", err)
+			}
+			s.attachmentIndices = append(s.attachmentIndices, idx)
+		case TokenChunkIndex:
+			idx, err := ParseChunkIndex(record)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse attachment index: %w", err)
+			}
+			s.chunkIndices = append(s.chunkIndices, idx)
+		case TokenStatistics:
+			stats, err := ParseStatistics(record)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse statistics: %w", err)
+			}
+			s.statistics = stats
+		case TokenFooter:
+			return s, nil
+		}
+	}
+}
+
+type recordIndex interface {
+	start() int64
+	len() int64
+}
+
+func (ci *ChunkIndex) start() int64      { return (int64)(ci.ChunkStartOffset) }
+func (ci *ChunkIndex) len() int64        { return (int64)(ci.ChunkLength + ci.MessageIndexLength) }
+func (ai *AttachmentIndex) start() int64 { return (int64)(ai.Offset) }
+func (ai *AttachmentIndex) len() int64   { return (int64)(ai.Length) }
+func (mi *MetadataIndex) start() int64   { return (int64)(mi.Offset) }
+func (mi *MetadataIndex) len() int64     { return (int64)(mi.Length) }
```

Comment:
I wonder if there's a better way we can handle this that would just result in failure in face of an out of range int64. FWIW uint64 support within the mcap tooling is a general weak spot. Even though the spec admits uint64 in many cases, all the golang stdlib functions for doing IO tend to take int64 and we haven't bothered to handle the accounting properly here, or I think on the TS side either.

Since the spec has uint64 seems right for our structs to contain them, but anywhere these functions lead to a downcast will probably end up breaking stuff.